### PR TITLE
write to shapefile

### PR DESCRIPTION
**Pull request by ErickChacon**
PR to be able to export to shapefile using `Shapefile.jl` and improve geointerface functionality. Whem geometries are `Multi`, exporting to `.shp` is quite easy with `SHP.write(fname, geotable)`, otherwise we need to convert to `Multi` and then write.

### CODE REVIEW

File: src/GeoTables.jl
Comment (juliohm) [inline src/GeoTables.jl:64]: Can you comment this code block? The `geotable` is a Tables.jl table with a column `:geometry`. Shapefile.jl doesn't support this format? Notice that MeshData(geotable)  shouldn't be necessary, and you can get the column of geometries with `geotable.geometry` as usual.

If all we need is convert non-Multi geometries to Multi before saving to disk, we can do so with:

```julia
newgeoms = [geom isa Multi ? geom : Multi([geom]) for geom in geotable.geometry]
newtable = georef(values(geotable), newgeomsi)
```

Comment (juliohm) [reply src/GeoTables.jl:64]: We could even avoid using `georef` and just use the GeoTable type to wrap a new table with the updated geometry column.

Comment (ErickChacon) [reply src/GeoTables.jl:64]: Yes, the main reason is that we need to convert to `Multi` with exception of `Points`. I will try your suggestion.

Comment (juliohm) [reply src/GeoTables.jl:64]: Is this a limitation of the shp file format or the Shapefile.jl implementation? Should this get fixed somewhere else instead of here?

File: src/conversion.jl
Comment (juliohm) [inline src/conversion.jl:23]: So the GI.ncoord must be defined for geometries too? I thought it was only defined for GI.PointTrait,at least that was the concept I remember from the last time I read the GI docs.

File: src/GeoTables.jl
Comment (ErickChacon) [reply src/GeoTables.jl:64]: Are you ok with something like the following?:
```julia   
newgeoms = [geom isa Multi ? geom : Multi([geom]) for geom in geotable.geometry]
meshdata(GeometrySet(newgeoms), etable = values(geotable))
```

File: src/GeoTables.jl
Comment (juliohm) [inline src/GeoTables.jl:34]: Awesome that you added kwargs, we should add this detail to the docstring of load and save.

File: src/GeoTables.jl
Comment (juliohm) [reply src/GeoTables.jl:64]: If we could avoid changing the geotable type before saving that would be ideal. I think the question we need to ask first is if this fix should go in Shapefile.jl instead.

File: src/conversion.jl
Comment (ErickChacon) [reply src/conversion.jl:23]: Yes, writing was not working because that was missing because the other packages use `GI.is3d` which depend of `GI.ncoord`. Hence, defining `GI.ncoord` helps with the conversion from `Meshes` to other packages to be able to write to other formats.

File: src/GeoTables.jl
Comment (ErickChacon) [reply src/GeoTables.jl:34]: Yes, it is quite useful. I needed that flexibility to show that we can read from `.geojson`, process with `Meshes.jl`, and then write to `.shp`. I can add some details.

File: src/GeoTables.jl
Comment (ErickChacon) [reply src/GeoTables.jl:64]: > Can you comment this code block? The `geotable` is a Tables.jl table with a column `:geometry`. Shapefile.jl doesn't support this format? Notice that MeshData(geotable) shouldn't be necessary, and you can get the column of geometries with `geotable.geometry` as usual.
> 
> If all we need is convert non-Multi geometries to Multi before saving to disk, we can do so with:
> 
> ```julia
> newgeoms = [geom isa Multi ? geom : Multi([geom]) for geom in geotable.geometry]
> newtable = georef(values(geotable), newgeomsi)
> ```

I think we should avoid redefining the data if not needed (Point and Multi), for this reason I used a conditional.

File: src/GeoTables.jl
Comment (juliohm) [inline src/GeoTables.jl:60]:
```suggestion
    if !(geoms[1] isa Multi) && !(geoms[1] isa Point)
```

File: src/GeoTables.jl
Comment (juliohm) [inline src/GeoTables.jl:63]: @ErickChacon I find it strange that we are fixing this issue in GeoTables.jl, shouldn't we contact the Shapefile.jl maintainers to fix this behavior over there? People will certainly try to save non-Multi geometries with Shapefile.jl and will get the same error right?

File: src/GeoTables.jl
Comment (ErickChacon) [reply src/GeoTables.jl]: Sorry, I missed some of your comments, it was weird that it was not showing before. I also thought if that should be a fix in `Shapefile.jl`, but I am not sure. It is not really a bug because `Shapefile.jl` read and write geometries as `Multi` with exception to `Points`, so they do not have (and do not need) a method to read or write a simple `LineString` or `Polygon`.

File: src/GeoTables.jl
Comment (juliohm) [reply src/GeoTables.jl:34]: It would be awesome to add some details to the docstring, thanks for these great improvements.

File: src/GeoTables.jl
Comment (ErickChacon) [reply src/GeoTables.jl:63]: Yes, that will also happen with other packages trying to use `Shapefile.jl`. I will start an issue to see if they can and want to handle that.

**Comment by ErickChacon** [comment]
@juliohm I started a discussion in `Shapefile.jl`, let's see what they say. I will also like to be able to write to other formats using `ArchGDAL`. Should I start that in a separate PR? I think we can use something similar to `GeoDataFrames.jl`:

https://github.com/evetion/GeoDataFrames.jl/blob/fec887c71124adf1dd11c33878ed6186aba10bd3/src/io.jl#L73C1-L164

Let me know your thoughts.

**Comment by juliohm** [comment]
Thank you @ErickChacon for these great additions. Can you explain why our save doesn't cover the ArchGDAL.jl case? If we pass a geotable with geometry column following the GeoInterface.jl shouldn't ArchGDAL.jl provide a similar AG.write function?

File: README.md
Comment (juliohm) [inline README.md:39]:
```suggestion
# read `.geojson` geometries with Float64 precision
table = GeoTables.load("file.geojson", numbertype = Float64)

# force writing on existing `.shp` file
GeoTables.save("file.shp", table, force = true)
```

File: src/GeoTables.jl
Comment (juliohm) [reply src/GeoTables.jl:34]: Thank you @ErickChacon for adding examples to the README. We also need to update the docstrings of load and save to reflect the new possibility of forwarding kwargs.

**Comment by ErickChacon** [comment]
> Thank you @ErickChacon for these great additions. Can you explain why our save doesn't cover the ArchGDAL.jl case? If we pass a geotable with geometry column following the GeoInterface.jl shouldn't ArchGDAL.jl provide a similar AG.write function?

Last time I checked it, there was no functionality to write from a table. I needed to manually convert a table to an `ArchGDAL` dataset iterating over rows to push `features`. I will check if this has been improved.

**Comment by juliohm** [comment]
That would be great. Maybe it is another situation where we need to fix the issue upstream in ArchGDAL.jl, adding a write function there before coming back here. This is a good exercise to improve the status of these loaders.

**Comment by ErickChacon** [comment]
> That would be great. Maybe it is another situation where we need to fix the issue upstream in ArchGDAL.jl, adding a write function there before coming back here. This is a good exercise to improve the status of these loaders.

Just to let you know that there is an attempt at yeesian/ArchGDAL.jl#243 to create an `IFeatureLayer` from a `Table`. Once something like that is merged, we could easily use `ArchGDAL` to export to other format types.

With respect to `Shapefile.jl`, I started an issue, should we accept our current fix until that issue is solved? Or should we keep the adequate code (removing the conditional lines) and accepting that an error will be thrown when attempting to write a `Chain` and `PolyArea` to a `.shp` file?

**Comment by juliohm** [comment]
Let's remove the conditional and wait for the Shapefile.jl fix then. I can
merge it here soon and release a new patch.

Em qua., 3 de mai. de 2023 20:40, Dr. Erick A. Chacón Montalván <
***@***.***> escreveu:

**Comment by juliohm** [comment]
Do we need more tests for the save functionality?

**Comment by juliohm** [comment]
I am wondering why our tests are failing only on GitHub Actions. Any clue why the HTTP requests are failing from GADM.jl?

**Comment by ErickChacon** [comment]
> Do we need more tests for the save functionality?

Yes, I forgot about it. Should I do it similar to the existing test with "geo.json"?

**Comment by ErickChacon** [comment]
> I am wondering why our tests are failing only on GitHub Actions. Any clue why the HTTP requests are failing from GADM.jl?

No, I haven't checked it. I will check it.

**Comment by juliohm** [comment]
> Should I do it similar to the existing test with "geojson"?

That would be great, but feel free to add other tests you have in mind too.

**Comment by juliohm** [comment]
I've updated the test/Project.toml in GADM.jl to see if the issue was coming from there, and learned that ArchGDAL.jl is not even building there in GitHub Actions: https://github.com/JuliaGeo/GADM.jl/issues/49

I really dislike the whole GDAL experience, so full of issues...

**Comment by ErickChacon** [comment]
> > Should I do it similar to the existing test with "geojson"?
> 
> That would be great, but feel free to add other tests you have in mind too.

Ok. Should it be better to make the `writing` test inside the formats `testset`, so we can take advantage of the already loaded data?

**Comment by juliohm** [comment]
Makes sense to move the writing tests to inside the other test sets. Maybe we can remove the outer "load" test set then?

**Comment by ErickChacon** [comment]
> Makes sense to move the writing tests to inside the other test sets. Maybe we can remove the outer "load" test set then?

Yes, that is what I have in mind. I will do that.

**Comment by ErickChacon** [comment]
> I've updated the test/Project.toml in GADM.jl to see if the issue was coming from there, and learned that ArchGDAL.jl is not even building there in GitHub Actions: [JuliaGeo/GADM.jl#49](https://github.com/JuliaGeo/GADM.jl/issues/49)

It builds in Julia 1.6 - ubuntu, but there is still an `HTTP` problem.

```
get: Error During Test at /home/runner/work/GADM.jl/GADM.jl/test/GADM.jl:34
  Got exception outside of a @test
  HTTP.Exceptions.RequestError(HTTP.Messages.Request:
  """
  GET /data/gadm3.6/gpkg/gadm36_SUR_gpkg.zip HTTP/1.1
  Host: biogeo.ucdavis.edu
  Accept: */*
  User-Agent: HTTP.jl/1.6.7
  Content-Length: 0
```

**Comment by ErickChacon** [comment]
I added tests for writing to `.shp` and `.geojson` for `points`, `lines`, `polygons` and `multipolygons`. I consider this PR completed.

File: test/runtests.jl
Comment (juliohm) [inline test/runtests.jl:205]: I think we should write to the datadir that is already defined. Can we avoid creating a new directory?

Comment (ErickChacon) [reply test/runtests.jl]: Done!

**Review by juliohm — APPROVED** [review]

**Comment by juliohm** [comment]
Thank you @ErickChacon! ❤️

**Pull request merged**
